experimental/StandardFiniteFields/StandardFiniteFields.jl

```diff
@@ -0,0 +1,462 @@
+    Fp = standard_finite_field(p,1)
+
+    get_steinitz_prime_degree!(Fp, r, k) do
+        # now we need to create the polynomial depending on the prime r
+        if r == p
+            # Artin-Schreier case
+            # k = 1 we get [(Xr[1])^p - (Xr[1]) -1]
+            # k > 1 we get (Xr[k])^p - (Xr[k]) - (prod(Xr[j] : j in [1..k-1]))^(p-1))
+            q = big(p)^(p^(k-1))
+            return (p-1)*(q + div(q,p))
+        elseif r == 2 && mod(p,4) == 3
+            if k == 1
+                # (Xr[1])^2 +1
+                return 1
+            elseif k == 2
+                a = non_rth_root(standard_finite_field(p,2), r)
+                # Xr[2]^2 -a
+                return steinitz_number(-a)
+            else
+                # Xr[i]^2 - Xr[i-1]
+                return (p-1)*big(p)^(r^(k-2))
+            end
+        elseif r == 2
+            if k == 1
+                # Xr[1]^2 -a
+                a = non_rth_root(standard_finite_field(p, 1), r)
+                return steinitz_number(-a)
+            else
+                # Xr[j]^r - Xr[j-1]
+                return (p-1)*big(p)^(r^(k-2))
+            end
+        else
+            # Here we use pseudo-random polynomials...
+            F = standard_finite_field(p, r^(k-1))
+            if k == 1
+                a = -one(F)
+            else
+                a = -gen(F)
+            end
+            l = standard_irreducible_coefficient_list(F,r,a)
+            pop!(l)
+            while is_zero(l[end])
+                pop!(l)
+            end
+            return evaluate(polynomial(ZZ, steinitz_number.(l)), order(F))
+        end
+    end
+end
+
+# x will be represented internally as a polynomial over Fp in the generator of F.
+# We need to first convert this to an Fp-vector, then to the corresponding vector
+# with respect to the Tower Basis.
+# Then we think of this vector as a polynomial (over ZZ) in a temporary indeterminate z,
+# and evaluate at z = char(F) to get the Steinitz number.
+# NOTE for whatever reason, evaluate(polynomial(), ) is faster than evalpoly()
+function steinitz_number(F::PrimeField, x::PrimeFieldElem)
+  @assert parent(x) === F
+  return lift(x)
+end
+function steinitz_number(F::FinField, x::FinFieldElem)
+    @assert parent(x) === F
+    v = lift.(absolute_coordinates(x) * primitive_powers_in_tower_basis(F))
+    return evaluate(polynomial(ZZ, v), characteristic(F))
+end
+function steinitz_number(x::FinFieldElem)
+    return steinitz_number(parent(x), x)
+end
+
+# describes monomials in tower basis plus their degrees
+function standard_monomial(n::IntegerUnion)
+    error("not implemented")
+end
+# just return degrees
+function standard_monomial_degrees(n::IntegerUnion)
+    if n == 1
+        return [1]
+    end
+    # need the largest prime factor a of n
+    nfactorization = factor(ZZ(n))
+    nfactors = sort([r for (r,e) in nfactorization])
+    a = Int(nfactors[end])
```

Comment:
You are allocating a vector, sorting it, just to compute the maximum. Better is to just ask for the maximum. I haven't tested it but perhaps something like this:
```suggestion
    a = maximum(x -> Int(x[1]), nfactorization)
```
Or alternatively even this:
```suggestion
    a = maximum(nfactorization)
```
at which point `a` is a pair, with `a[1]` resp. `a.first` taking the place of the old `a`, and `a[2]` resp. `a.second` replacing `nfactorization[a]`.

Reply:
I've made this change, but the more I think about it, this is used in two recursive functions where I factor n, divide out the largest prime factor to obtain n', and then pass this to a new function call that.... factor it again. So I will work on passing the factorization through instead of repeating the work.